### PR TITLE
Adds `{ModelRequest,ModelResponse}.metadata` fields

### LINKED ISSUE #3404: Add metadata field to ModelMessage
### Summary

This feature request is to add a new, optional `metadata: dict[str, Any] | None` field to both the `ModelRequest` and `ModelResponse`. This field would serve as a "user-space" container for arbitrary, application-specific data that is not sent to the LLM.

### Problem & Motivation

Developers often need to tag messages for filtering, grouping, or post-processing based on their application's internal logic. However, currently, there is no dedicated field on message objects to attach custom, application-specific data.

For example, an application might need to decorate the user's raw input before sending it as a `ModelRequest`. With a `metadata` field, developers could easily track which messages are customized and which template was used, enabling more reliable filtering or post-processing. Without metadata, we currently have to rely on detecting special characters in the message, which is error-prone.

A dedicated `metadata` field provides a clean, and simple way to store this information directly on the message object, improving developer experience and code robustness. 

### References

_No response_

### PR DESCRIPTION
**Pull request by adtyavrdhn**
Closes #3404 

Adds metadata field to ModelMessage.

### CODE REVIEW

**Comment by adtyavrdhn** [comment]
I've not added a history processor test for the same because I did not think it was relevant to the field or history processor itself, but if you think we should add one as a valid use case scenario let me know :)

File: pydantic_ai_slim/pydantic_ai/messages.py [hunk 1]
```diff
@@ -970,6 +970,9 @@ class ModelRequest:
     run_id: str | None = None
     """The unique identifier of the agent run in which this message originated."""
 
+    metadata: dict[str, Any] | None = None
+    """Additional data tagged on the message but is not sent to the model"""
```
Comment (DouweM) [inline pydantic_ai_slim/pydantic_ai/messages.py]: I like the description from `ToolReturn.metadata`:

```suggestion
    """Additional data that can be accessed programmatically by the application but is not sent to the LLM."""
```

**Comment by adtyavrdhn** [comment]
## Code Review

I've reviewed the implementation of the metadata field addition to ModelRequest and ModelResponse. The overall implementation is solid and follows the existing codebase patterns well. Here are a few suggestions for improvement:

### Issues Found

**1. Docstring Consistency (messages.py:1095, 1393)**
The metadata field docstring is identical in both ModelRequest and ModelResponse, but the context differs. Consider clarifying:
- For ModelRequest: metadata attached by the application before sending to LLM
- For ModelResponse: metadata related to response processing or analysis

**2. Test Coverage and Naming (tests/test_messages.py:557)**
The test  was modified to test metadata but the name still focuses on run_id. Either:
- Rename to reflect it tests both run_id and metadata preservation, or  
- Add dedicated tests specifically for metadata functionality

**3. Missing Test Coverage**
The changes lack comprehensive test coverage for metadata functionality. Consider adding tests for:
- Metadata handling in isolation
- metadata=None behavior
- Verification that metadata doesn't affect LLM interaction
- Various data types in metadata as expected for dict[str, Any]

**4. Documentation Gap (docs/message-history.md)**
The new metadata field should be documented in the message history documentation to show users how to attach and access custom metadata, especially given the use cases mentioned in issue #3404 (template tracking, message filtering).

### Positive Notes

- ✅ Type annotation follows existing patterns ()
- ✅ Field placement and structure consistent with codebase
- ✅ Backwards compatibility maintained with None default
- ✅ Serialization/deserialization works correctly

The core implementation is good - these suggestions focus on improving documentation, test coverage, and developer experience.

**Comment by adtyavrdhn** [comment]
## Code Review for PR #3422: Add ThinkingPartDelta

This PR adds streaming support for thinking parts by introducing  class. The implementation is functionally sound but has some issues that should be addressed:

### Issues Found

1. **Missing docstring for ThinkingPartDelta class** (, line ~1651)
   - All public API classes should have docstrings explaining their purpose and usage
   - Should document what delta application means for thinking parts

2. **Missing type annotation** (, line ~1652) 
   - The  field lacks an explicit type annotation
   - Should be typed as  for clarity

3. **Insufficient test coverage** (, line ~365)
   - Only basic instantiation test exists
   - Missing tests for: None thinking field handling, delta application behavior, integration scenarios
   - Consider adding tests that verify the streaming delta application works correctly

### Summary

The core implementation is correct and follows the existing patterns well. The new class is properly exported in . However, better documentation and more comprehensive testing would improve the robustness of this public API addition.

**Comment by adtyavrdhn** [comment]
## Code Review for PR #3422

Thank you for implementing the metadata fields for ModelRequest and ModelResponse! This addresses issue #3404 well. I've identified a few areas for improvement:

### Type Safety Concerns
**File:  (lines 1096, 1394)**
The  type annotation is quite permissive and could lead to type safety issues. Consider:
- Using a more constrained type like  for common serializable types
- Or creating a custom TypedDict for expected metadata structure
- This would provide better IDE support and catch type errors at development time

### Test Coverage
**File:  (lines 873-884)**
The current tests only cover basic assignment and retrieval. Consider adding tests for:
- Empty metadata dict behavior
- Nested data structures in metadata
- Serialization/deserialization behavior (if applicable)
- Integration with actual agent workflow to ensure metadata persists correctly
- Edge cases with None values

### Documentation
**File:  (lines 1095, 1393)**
The docstring could be more comprehensive:
- Provide concrete examples of appropriate metadata usage
- Clarify the lifecycle - when metadata is preserved vs. lost during agent operations  
- Specify any constraints on the data types or structure
- Example: """Additional application data (e.g., request IDs, user context, debugging info) that persists through the agent workflow but is not sent to the LLM."""

The core implementation looks solid and the feature will be valuable for tracking additional context through agent interactions!
